geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/stats/GeoMesaMetadataStats.scala

```diff
@@ -0,0 +1,479 @@
+          sft.getDescriptor(attribute).getType.getBinding match {
+            case c if c.isAssignableFrom(classOf[String])   => ("0", "z")
+            case c if c.isAssignableFrom(classOf[Integer])  => (Int.box(0), Int.box(1000))
+            case c if c.isAssignableFrom(classOf[jLong])    => (Long.box(0L), Long.box(1000L))
+            case c if c.isAssignableFrom(classOf[jFloat])   => (Float.box(0f), Float.box(1000f))
+            case c if c.isAssignableFrom(classOf[jDouble])  => (Double.box(0.0), Double.box(1000.0))
+            case c if c.isAssignableFrom(classOf[Date])     => (now.minusWeeks(1).toDate, now.toDate)
+            case c => throw new NotImplementedError(s"Can't handle bounds of type $c")
+          }
+        }
+
+        // note: we need the class-tag binding to correctly create the stat
+        bounds match {
+          case (s: String,   e: String)   => Stat.RangeHistogram(attribute, DefaultHistogramSize, s, e)
+          case (s: Integer,  e: Integer)  => Stat.RangeHistogram(attribute, DefaultHistogramSize, s, e)
+          case (s: jLong,    e: jLong)    => Stat.RangeHistogram(attribute, DefaultHistogramSize, s, e)
+          case (s: jFloat,   e: jFloat)   => Stat.RangeHistogram(attribute, DefaultHistogramSize, s, e)
+          case (s: jDouble,  e: jDouble)  => Stat.RangeHistogram(attribute, DefaultHistogramSize, s, e)
+          case (s: Date,     e: Date)     => Stat.RangeHistogram(attribute, DefaultHistogramSize, s, e)
+          case b => throw new NotImplementedError(s"Can't handle bounds $b")
+        }
+      }
+    }
+
+    Stat.SeqStat(Seq(count) ++ minMax ++ histograms)
+  }
+}
+
+/**
+  * Stores stats as metadata entries
+  *
+  * @param stats persistence
+  * @param sft simple feature type
+  * @param statFunction creates stats for tracking new features - this will be re-created on flush,
+  *                     so that our bounds are more accurate
+  */
+class MetadataStatUpdater(stats: GeoMesaMetadataStats, sft: SimpleFeatureType, statFunction: => Stat)
+    extends StatUpdater with LazyLogging {
+
+  private var stat: Stat = statFunction
+
+  override def add(sf: SimpleFeature): Unit = stat.observe(sf)
+
+  override def remove(sf: SimpleFeature): Unit = stat.unobserve(sf)
+
+  override def close(): Unit = {
+    stats.writeStat(stat, sft, merge = true)
+    // schedule a compaction so our metadata doesn't stack up too much
+    stats.ds.connector.tableOperations().compact(stats.ds.catalogTable, null, null, true, false)
+  }
+
+  override def flush(): Unit = {
+    stats.writeStat(stat, sft, merge = true)
+    // reload the tracker - for long-held updaters, this will refresh the histogram ranges
+    stat = statFunction
+  }
+}
+
+object GeoMesaMetadataStats {
+
+  // how many buckets to sort each attribute into
+  // max space on disk = 8 bytes * size - although we use optimized serialization so likely 1-3 bytes * size
+  val DefaultHistogramSize  = 1000
```

Comment:
I need to re-evaluate this - once we get up to 100's of millions of points, the histograms will be mostly worthless. Probably need to grow them dynamically or split them up by date range or something.